### ISSUE TITLE
Update Losses to Gen 2

### ISSUE THREAD

**Issue by djgagne** [NCAR/miles-credit#402]
Losses need to be updated to accept Gen 2 state dict, handle `computed_diagnostics` variables. MSE definitely should follow this framework. Other losses could potentially be added in 2026.2.1.

**Comment by djgagne** [comment]
Currently the call to the loss in `trainer_gen2` takes y_pred as input,which is the raw output from the model.
https://github.com/NCAR/miles-credit/blob/595cf700f1fb415d035263687f032f92c153d327/credit/trainers/trainer_gen2.py#L175-L186,  The loss should be calculated against `full_data_dict["y_processed"]`, and we need to make sure that all of the post blocks are updating the state in `full_data_dict["y_processed"]`, which may not be happening with geopotential and mslp at the moment. 

There is also an open question about how to process the target data for the model. Does it need to go through both the preblocks and postblocks? Does it need to be renormalized before the loss or have some kind of normalization or variance/covariance as part of the loss?

**Comment by djgagne** [comment]
Update from discussion:
* Update gen 2 trainer to use gen 2 loss call that accepts `y_processed` and target (or y) as arguments along with variable weights.
* Add BaseLoss class that handles
  * check that keys match in pred and target dicts.
  * loop or vmap through calculations on each variable
  * call loss functions
  * aggregate everything

**Comment by djgagne** [comment]
I have a draft BaseLoss class in the `g2loss` branch.  See linked development section. It passes tests but I want to do a more complete evaluation. I plan to test on casper and derecho first and then will submit the PR.

**Comment by danineamati** [comment]
Following up on today's meeting, I took a look at `BaseLoss`. I am on board with the Kendall-Gal uncertainty as an option. I wonder if we want to cite the paper, which I believe is [https://arxiv.org/abs/1705.07115](https://arxiv.org/abs/1705.07115). I believe it is quite similar, though distinct, from the loss I was thinking of: [`torch.nn.GaussianNLLLoss`](https://docs.pytorch.org/docs/2.13/generated/torch.nn.GaussianNLLLoss.html).

To be more mathematical, Kendall _et al._ has equation (7):

$$\mathcal{L} = \frac{1}{2 \sigma_1^2} \mathcal{L}_1(W) +\frac{1}{2 \sigma_2^2} \mathcal{L}_2(W) + \log (\sigma_1 \sigma_2) $$

for $\mathcal{L}_1(W)$ and $\mathcal{L}_2(W)$  of the form $\| y - f_W(x) \|^2 $. 

Whereas we have:
https://github.com/NCAR/miles-credit/blob/eebefa9c145fb43366bee77f52db113b2b728246/credit/losses/base.py#L426-L432

Which simplifies as:

$$\mathcal{L} = w_{\text{manual}} \exp(-\log(\sigma^2)) \mathcal{L}_{\text{variable}} + \frac{1}{2} \log(\sigma^2) $$

$$= \frac{w_{\text{manual}}}{\sigma^2} \mathcal{L}_{\text{variable}} + \log(\sigma) $$

and our $$\mathcal{L}_{\text{variable}}$$ could be MSE among other options.

As a side note, I think "var" becomes ambiguous in parts of the code since it could refer to "variance" or "variable."
